jep/0001/README.adoc

```diff
@@ -0,0 +1,306 @@
+
+==== Run metadata
+
+The following metadata entries are required to be present
+for all `hudson.model.Run` event entries:
+
+* All event metadata
+* `jobId` - ID of the Job, should be provided by Unique ID Plugin
+
+==== Pipeline metadata
+
+* All Run metadata
+* `stepId` - Identifier of the step ("node")
+
+=== Extension Points
+
+The plugin should offer the following extension points:
+
+* `ExternalLoggingMethodFactory` -
+produced `ExternalLoggingMethod` instances for Runs and, eventually, other objects.
+This extension point implements `Describable`
+and should be configurable via UI.
+* `ExternalLogBrowserFactory` -
+Same as above, but for `ExternalLogBrowser`
+
+=== ExternalLoggingMethodFactory and classes
+
+Produces `ExternalLoggingMethod` instances for Runs and, eventually, other objects.
+This extension point implements `Describable`
+and should be configurable via UI.
+
+ExternalLoggingMethod is a new extension point on the top of `LoggingMethod`:
+
+* It implements `LoggingMethod` interfaces
+* It provides a new `createWriter()` method,
+which produces a `ExternalLoggingEventWriter` class instance.
+** This method automatically injects the mandatory metadata for events
+** An abstract `_createWriter()` method is offered for implementation by downstream implementations
+* Being compared to `LoggingMethod`,
+the implementation is designed to always perform logging
+from the agent side
+
+ExternalLoggingEventWriter class:
+
+* The class is `Serializable`.
+It will be sent to the agent side.
+* The class offers the following abstract methods:
+** `void writeEvent(Event event) throws IOException`
+- writing event to the remote storage
+* The class also stores metadata, which may be injected into the events
+** The class stores a Map of Serializable metadata entries
+** The class offers API, which allow setting the metadata.
+   This API will be used by `ExternalLoggingMethod` implementations and other logic
+   to provide additional metadata if required
+
+=== ExternalLogBrowserFactory and classes
+
+This factory just produces instances of `ExternalLogBrowser`.
+This class offers an abstraction layer for external log browsing.
+
+`ExternalLogBrowser` will also provide an abstraction layer for
+eventual consistency management.
+This layer will be determined during reference implementation polishing.
+
+=== Pipeline Bridge extensions
+
+The plugin should also implement Pipeline `LogStorage` and `LogStorageFactory` extension points
+so that it transparently supports Pipeline with existing API.
+
+Pipeline Storage JEP is documented in JEP-TODO.
+
+=== JSON-based external-logging layer
+
+The API Plugin should offer a convenience layer in order
+to support a number of most common logging providers like
+Logstash, Fluentd, Elasticsearch, AWS CloudWatch, etc.
+
+This layer should provide the following features:
+
+* Utility classes for reading and writing JSON Events,
+including parsing/writing `ConsoleNote` objects
```

Comment:
This is not really specific to JSON. Would apply to any log sink accepting structured records of any format.